src/shop.c

```diff
@@ -0,0 +1,1283 @@
+    {gText_ShopQuit, {.void_u8 = Task_HandleShopMenuQuit}}
+};
+
+static const struct YesNoFuncTable sShopMenuActions_BuyQuit[] =
+{
+    BuyMenuTryMakePurchase,
+    BuyMenuReturnToItemList
+};
+
+static const struct WindowTemplate sShopMenuWindowTemplate =
+{
+    .bg = 0,
+    .tilemapLeft = 2,
+    .tilemapTop = 1,
+    .width = 12,
+    .height = 6,
+    .paletteNum = 15,
+    .baseBlock = 8
+};    
+
+static const struct BgTemplate sShopBuyMenuBgTemplates[4] =
+{
+    {
+        .bg = 0,
+        .charBaseIndex = 2,
+        .mapBaseIndex = 31,
+        .screenSize = 0,
+        .paletteMode = 0,
+        .priority = 0,
+        .baseTile = 0
+    },
+    {
+        .bg = 1,
+        .charBaseIndex = 0,
+        .mapBaseIndex = 30,
+        .screenSize = 0,
+        .paletteMode = 0,
+        .priority = 1,
+        .baseTile = 0
+    },
+    {
+        .bg = 2,
+        .charBaseIndex = 0,
+        .mapBaseIndex = 29,
+        .screenSize = 0,
+        .paletteMode = 0,
+        .priority = 2,
+        .baseTile = 0
+    },
+    {
+        .bg = 3,
+        .charBaseIndex = 0,
+        .mapBaseIndex = 28,
+        .screenSize = 0,
+        .paletteMode = 0,
+        .priority = 3,
+        .baseTile = 0
+    }
+};
+
+// Functions
+static u8 CreateShopMenu(u8 a0)
+{
+    gShopData.martType = GetMartTypeFromItemList(a0) & 0xF;
+    gShopData.selectedRow = 0;
+    if (ContextNpcGetTextColor() == 0)
+        gShopData.unk16_4 = 4;
+    else
+        gShopData.unk16_4 = 5;
+    
+    sShopMenuWindowId = AddWindow(&sShopMenuWindowTemplate);
+    SetStdWindowBorderStyle(sShopMenuWindowId, 0);
+    PrintTextArray(sShopMenuWindowId, 2, GetMenuCursorDimensionByFont(2, 0), 2, 16, 3, sShopMenuActions_BuySellQuit);
+    Menu_InitCursor(sShopMenuWindowId, 2, 0, 2, 16, 3, 0);
+    PutWindowTilemap(sShopMenuWindowId);
+    CopyWindowToVram(sShopMenuWindowId, 1);
+    return CreateTask(Task_ShopMenu, 8);
+}
+
+static bool8 GetMartTypeFromItemList(bool32 a0)
+{    
+    u16 i;
+    
+    if (a0)
+        return (u8)a0;
+    goto MAIN;
+    
+    RETURN_1:
+    return 1;
+
+    MAIN:
+    i = 0;
+    if (i >= gShopData.itemCount)
+        goto RETURN_0;
+    else if (gShopData.itemList[0] == 0)
+        goto RETURN_0;
+
+    while (gShopData.itemList[i] != 0)
+    {
+        if (ItemId_GetPocket(gShopData.itemList[i]) == POCKET_TM_CASE)
+            goto RETURN_1;
+        ++i;
+        if (i >= gShopData.itemCount)
+            goto RETURN_0;
+        else if (gShopData.itemList[i] == 0)
+            goto RETURN_0;
+    }
+    
+    RETURN_0:
+    return 0;
+}
+
+static void SetShopItemsForSale(const u16 *items)
+{    
+    struct ShopData *mart;
+
+    mart = &gShopData;
+    mart->itemList = items;
+    mart->itemCount = 0;
+    if (mart->itemList[0] == 0)
+        return;
+
+    do
+    {
+        ++gShopData.itemCount;
+    } while (mart->itemList[gShopData.itemCount]);
+}
+
+static void SetShopMenuCallback(MainCallback callback)
```

Comment:
I don't think this is `MainCallback`. It's called in this file's function instead, so either typedef another identifier or simply write `void (*callback)(void)`